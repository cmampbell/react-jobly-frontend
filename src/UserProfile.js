import React, { useState, useEffect } from 'react';
import { useLoaderData, Form, useOutletContext } from 'react-router-dom';

const UserProfile = () => {
    const user = useLoaderData();
    const { editUser, error, setError, setMessage } = useOutletContext()
    const INITIAL_STATE = {
        username: user.username,
        firstName: user.firstName,
        lastName: user.lastName,
        email: user.email
    };
    const [formData, setFormData] = useState(INITIAL_STATE);


    const handleChange = (evt) => {
        const { name, value } = evt.target
        setFormData(() => ({
            ...formData, [name]: value
        }
        ))
    }

    const handleSubmit = (evt) => {
        evt.preventDefault()
        const user = editUser(formData)
        setMessage(`Update successful for ${user.username}`)
    }

    // clear error and message after component unmounts
    useEffect(() => () => {
        setMessage('')
        setError('')
    }, [setError, setMessage])

    return (
        <Form onSubmit={handleSubmit}>
            <label htmlFor='username'>Username:</label>
            <input
                name="username"
                id="username"
                type="text"
                readOnly
                value={formData.username}
                onChange={handleChange} />
            <label htmlFor='firstName'>First name:</label>
            <input
                name="firstName"
                id="firstName"
                type="text"
                value={formData.firstName}
                onChange={handleChange} />
            <label htmlFor='lastName'>Last name:</label>
            <input
                name="lastName"
                id="lastName"
                type="text"
                value={formData.lastName}
                onChange={handleChange} />
            <label htmlFor='email'>Email:</label>
            <input
                name="email"
                id="email"
                type="text"
                value={formData.email}
                onChange={handleChange} />
            {error && <p>Unable to Update{error.message}</p>}
            <button type="submit">Edit {user.username}!</button>
        </Form>
    )
}

export default UserProfile;