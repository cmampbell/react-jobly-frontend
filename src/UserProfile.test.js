import { render } from '@testing-library/react';
import UserProfile from './UserProfile';
import { RouterProvider, createMemoryRouter} from 'react-router-dom';

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useOutletContext: function () {
        return {
            editUser: () => null,
            error: null,
            setError: ()=> null,
            setMessage: ()=> null
        }
    }
})
);

const testUser = {
    username: 'test',
    firstName: 'testFN',
    lastName: 'testLN',
    email: 'test@test.com'
}

const route = [
    {
      path: "/",
      element: <UserProfile />,
      loader: () => testUser
    },
  ];

const router = createMemoryRouter(route);

describe('Smoke and snapshot CompanyCard test', () => {
    it('should render without crashing', () => {
        render(<RouterProvider router={router}/>);
    });
    it('should match snapshot', async ()=> {
        const { asFragment, findByText } = render(<RouterProvider router={router}/>);
        expect(asFragment()).toMatchSnapshot();
        expect(await findByText('Edit test!')).toBeInTheDocument()
    })
})