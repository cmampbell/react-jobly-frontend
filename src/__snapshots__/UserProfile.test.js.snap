// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke and snapshot CompanyCard test should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="/"
    method="get"
  >
    <label
      for="username"
    >
      Username:
    </label>
    <input
      id="username"
      name="username"
      readonly=""
      type="text"
      value="test"
    />
    <label
      for="firstName"
    >
      First name:
    </label>
    <input
      id="firstName"
      name="firstName"
      type="text"
      value="testFN"
    />
    <label
      for="lastName"
    >
      Last name:
    </label>
    <input
      id="lastName"
      name="lastName"
      type="text"
      value="testLN"
    />
    <label
      for="email"
    >
      Email:
    </label>
    <input
      id="email"
      name="email"
      type="text"
      value="test@test.com"
    />
    <button
      type="submit"
    >
      Edit test!
    </button>
  </form>
</DocumentFragment>
`;
